src/Servers/IIS/IIS/src/Core/IISHttpContext.cs

```diff
@@ -615,10 +615,24 @@ protected void ReportApplicationError(Exception ex)
         Log.ApplicationError(_logger, ((IHttpConnectionFeature)this).ConnectionId, ((IHttpRequestIdentifierFeature)this).TraceIdentifier, ex);
     }
 
+    // This method should only be called from the end of request processing, this lets use avoid most
+    // of the rentrancy issues with using indicate completion.
     public void PostCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS requestNotificationStatus)
     {
-        NativeMethods.HttpSetCompletionStatus(_requestNativeHandle, requestNotificationStatus);
-        NativeMethods.HttpPostCompletion(_requestNativeHandle, 0);
+        // We post completion to the thread pool instead of the IIS thread pool to avoid having
+        // new requests compete with existing threads for IIS threads.
+        // We're going to use the local queue so that we can resume the state machine on this thread after
+        // unwinding.
+        ThreadPool.UnsafeQueueUserWorkItem(static state =>
+        {
+            var (handle, status) = state;
+
+            // This will resume the IIS state machine inline
+            NativeMethods.HttpIndicateCompletion(handle, status);
+        },
+        // We use the native handle because we're going to dispose the safe handle by the time this runs.
+        // We don't want to hold onto it since managed code is unwinding.
+        (_requestNativeHandle.DangerousGetHandle(), requestNotificationStatus), preferLocal: true);
```

Comment:
So this is safe to use inside the callback because we await the VTS right after calling PostCompletion? No one else can clean up the handle beneath us? Or could a server shutdown clean up the handle while the work item is queued?

Reply:
> So this is safe to use inside the callback because we await the VTS right after calling PostCompletion?

The VTS only protects the safe handle. This is the native handle, the only thing that cleans this up is native code. So when IIS itself is cleaning up.

This comment made me look at the native code though, I need to change it to set `IndicateManagedRequestComplete` in indicate completion.

Reply:
OK this made me look around:
- The native code checks if we're done with the managed request. I made a change to the code to set that boolean before we run this logic so it's similar to before.
- We have a new native -> managed code dependency to complete requests. I think that means if there's a native module blocking a thread, it'll end up blocking a thread pool thread. This didn't happen before as we were posting completions back to IIS before resuming the state machine.
- I guess it also means when shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads so AV?

Reply:
>  shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads

I have a few issues assigned to me about our Stopping events not reliably firing, is this possibly related?

Reply:
I don't know if we can do anything useful in this case besides wait a little bit. Also testing it would require writing a misbehaving native module and blocking in it, then shutting down. I debugged the codez last night and have a little better understanding of the flow.

Reply:
> I guess it also means when shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads so AV? 

I think what we want to do is to call `_requestNativeHandle.Dispose()` and await `await new ValueTask<object?>(_requestNativeHandle, _requestNativeHandle.Version);` before calling `CompleteHttpRequest()`. It was previously `http_post_completion` and now presumably `http_indicate_completion` that tells IIS that it's okay to release the backing handle.

The reason we have the weird dispose the SafeHandle then await it logic is to ensure any concurrent p/invokes with the SafeHandle have completed and any future ones will ODE. I think the logic was wrong here previously. We should have been doing this before calling `PostCompletion()` too.

Fortunately, this really only affects buggy code that tries to use the `HttpContext` after middleware complets. Still, it's better to avoid the possible AV.

Reply:
> I think what we want to do is to call _requestNativeHandle.Dispose() and await await new ValueTask<object?>(_requestNativeHandle, _requestNativeHandle.Version); before calling CompleteHttpRequest(). It was previously PostCompletion() and now presumably IndicateCompletion() that tells IIS that it's okay to release the backing handle.

The safe handle doesnt do much and isn't useful here. Indicate completion literally runs the rest of the request on the thread that calls it.

Reply:
We should probably just update `NativeSafeHandle` to call `NativeMethods.HttpIndicateCompletion(handle, status)` and drop the `IValueTaskSource` stuff. You'd have to add a property or something for the status.

Then we could just dispatch the disposal of the `NativeSafeHandle`.